File: internal/drill/drill.go
// Package drill は練習セッションの進行と昇格・降格の判定を行う。
//
// 出題は1単語ずつ。単語が表示された瞬間から打ち終わるまでを計測する。
// 昇格は直近の窓の打鍵速度 (kpm) とミス率の両方が基準を満たしたとき。
// 並行安全ではない。
package drill

import (
	"slices"
	"time"

	"github.com/ktnyt/shihandai/internal/curriculum"
	"github.com/ktnyt/shihandai/internal/naginata"
)

// recentWindow は1つのかなの直近正答率を測る試行数。
const recentWindow = 12

// UnitStat は1つのかなの成績。
type UnitStat struct {
	Attempts int    `json:"attempts"`
	Errors   int    `json:"errors"`
	Recent   []bool `json:"recent"` // 直近の試行の成否
}

func (s *UnitStat) record(ok bool) {
	s.Attempts++
	if !ok {
		s.Errors++
	}
	s.Recent = append(s.Recent, ok)
	if len(s.Recent) > recentWindow {
		s.Recent = s.Recent[len(s.Recent)-recentWindow:]
	}
}

// RecentAccuracy は直近の正答率を返す。試行がなければ1を返す。
func (s *UnitStat) RecentAccuracy() float64 {
	if len(s.Recent) == 0 {
		return 1
	}
	ok := 0
	for _, r := range s.Recent {
		if r {
			ok++
		}
	}
	return float64(ok) / float64(len(s.Recent))
}

// Config は判定の調整項目。
type Config struct {
	TargetKPM       float64       // 昇格に必要な打鍵速度 (keys per minute)
	MaxMissRate     float64       // 昇格できるミス率の上限
	ReactionBudget  time.Duration // 表示から打ち始めるまでの猶予。kpm の計算で引く
	WindowSize      int           // 判定に使う直近の単語数
	MinNewKanaWords int           // 昇格までに打つ、新出かなを含む語の最低数
	DemoteAccuracy  float64       // これを下回ると降格するかなの直近正答率
	MinAttempts     int           // 降格判定に必要な直近試行数
}

// DefaultConfig は既定値を返す。
// 降格はゆるめにしてある。新しいかなが入ると既存のかなも文脈が変わって
// 一時的に難しくなるので、直近12回が埋まった上で4ミス以上（70%未満）で
// はじめて降格する。
func DefaultConfig() Config {
	return Config{
		TargetKPM:       120,
		MaxMissRate:     0.05,
		ReactionBudget:  500 * time.Millisecond,
		WindowSize:      100,
		MinNewKanaWords: 50,
		DemoteAccuracy:  0.70,
		MinAttempts:     recentWindow,
	}
}

// WordRecord は判定の窓に残る、1単語分の記録。
type WordRecord struct {
	Success bool          `json:"success"`
	Units   int           `json:"units"`  // 正しく打ったかなの数
	Keys    int           `json:"keys"`   // 打鍵数 (同時押しは複数と数える)
	Errors  int           `json:"errors"` // ミス入力の数
	Typing  time.Duration `json:"typing"` // 反応の猶予を引いた打鍵時間
}

// Drill は練習全体の状態。
type Drill struct {
	Cfg   Config
	Level int
	Stats map[string]*UnitStat

	word          []string
	pos           int
	wordErrors    int
	shownAt       time.Time
	records       []WordRecord // 直近 WindowSize 単語の記録
	newKanaWords  int          // このレベルで打った、新出かなを含む語の数
	newKanaSupply int          // 新出かなを含む語が辞書に何語あるか。負なら不明
}

// New は練習状態を作る。
func New(cfg Config, level int, stats map[string]*UnitStat) *Drill {
	if stats == nil {
		stats = map[string]*UnitStat{}
	}
	if level < 1 {
		level = 1
	}
	if level > curriculum.MaxLevel() {
		level = curriculum.MaxLevel()
	}
	return &Drill{Cfg: cfg, Level: level, Stats: stats, newKanaSupply: -1}
}

// Allowed は現在のレベルで使えるかなを返す。
func (d *Drill) Allowed() []string { return curriculum.For(d.Level) }

// Stage は現在のレベルの出題条件を返す。
func (d *Drill) Stage() curriculum.Stage { return curriculum.StageFor(d.Level) }

// Newest はいちばん新しく増えたかなを返す。
func (d *Drill) Newest() string {
	allowed := d.Allowed()
	return allowed[len(allowed)-1]
}

// NewKanaWords はこのレベルで打った、新出かなを含む語の数を返す。
func (d *Drill) NewKanaWords() int { return d.newKanaWords }

// SetNewKanaSupply は新出かなを含む語が辞書に何語あるかを教える。
// 語彙の少ないかなでゲートが満たせなくなるのを防ぐのに使う。
func (d *Drill) SetNewKanaSupply(n int) { d.newKanaSupply = n }

// supplyFactor は語彙の少ないかなのゲートの倍率。
// 辞書に4語しかなければ 4×5=20 語打てば足りる。
const supplyFactor = 5

// GateTarget は昇格までに打つべき、新出かなを含む語の数を返す。
// 最初の5文字の段階では全部が新出なので、ゲートはかけない。
func (d *Drill) GateTarget() int {
	if len(d.Allowed()) == len(curriculum.For(1)) {
		return 0
	}
	target := d.Cfg.MinNewKanaWords
	if d.newKanaSupply >= 0 {
		target = min(target, d.newKanaSupply*supplyFactor)
	}
	return target
}

// Progress は保存のために判定の窓とゲートのカウンタを返す。
func (d *Drill) Progress() (records []WordRecord, newKanaWords int) {
	return d.records, d.newKanaWords
}

// SetProgress は保存されていた進捗を復元する。
func (d *Drill) SetProgress(records []WordRecord, newKanaWords int) {
	if len(records) > d.Cfg.WindowSize {
		records = records[len(records)-d.Cfg.WindowSize:]
	}
	d.records = records
	d.newKanaWords = max(newKanaWords, 0)
}

// StartWord は新しい単語を出題する。now は表示した時刻。
func (d *Drill) StartWord(units []string, now time.Time) {
	d.word = units
	d.pos = 0
	d.wordErrors = 0
	d.shownAt = now
}

// Word は現在の単語を返す。
func (d *Drill) Word() []string { return d.word }

// Pos は現在の入力位置を返す。
func (d *Drill) Pos() int { return d.pos }

// WordErrors は現在の単語でのミス数を返す。
func (d *Drill) WordErrors() int { return d.wordErrors }

// Expected は次に打つべきかなを返す。単語の終わりなら空文字。
func (d *Drill) Expected() string {
	if d.pos >= len(d.word) {
		return ""
	}
	return d.word[d.pos]
}

// Elapsed は表示からの経過時間を返す。
func (d *Drill) Elapsed(now time.Time) time.Duration {
	if d.shownAt.IsZero() {
		return 0
	}
	return now.Sub(d.shownAt)
}

// wordKeys は現在の単語の打鍵数を返す。
func (d *Drill) wordKeys() int {
	keys := 0
	for _, u := range d.word {
		if chord, ok := naginata.ChordFor(u); ok {
			keys += chord.Count()
		} else {
			keys++
		}
	}
	return keys
}

// Result は1回の入力の判定結果。
type Result int

const (
	// ResultIgnored は判定対象外の入力（空白、制御文字など）。
	ResultIgnored Result = iota
	// ResultAdvance は正しい入力で1単位進んだ。
	ResultAdvance
	// ResultError は誤った入力。位置は進まない。
	ResultError
	// ResultWordDone は単語の最後の単位を正しく入力した。
	ResultWordDone
)

// Input は確定した1単位を判定する。
func (d *Drill) Input(text string) Result {
	if d.pos >= len(d.word) {
		return ResultIgnored
	}
	switch text {
	case " ", "\b", "\n", "":
		return ResultIgnored
	}

	expected := d.word[d.pos]
	if text == expected {
		d.stat(expected).record(true)
		d.pos++
		if d.pos >= len(d.word) {
			return ResultWordDone
		}
		return ResultAdvance
	}
	d.stat(expected).record(false)
	d.wordErrors++
	return ResultError
}

func (d *Drill) stat(unit string) *UnitStat {
	// 手で編集された進捗ファイル由来の nil 値にも耐える
	s := d.Stats[unit]
	if s == nil {
		s = &UnitStat{}
		d.Stats[unit] = s
	}
	return s
}

// SuccessCount は直近の成功数と試行数を返す。
func (d *Drill) SuccessCount() (successes, total int) {
	for _, r := range d.records {
		if r.Success {
			successes++
		}
	}
	return successes, len(d.records)
}

// WindowKPM は直近の窓の打鍵速度 (keys per minute) を返す。
// 各単語の経過時間から反応の猶予を引いた分を打鍵時間とみなす。
func (d *Drill) WindowKPM() float64 {
	keys := 0
	var typing time.Duration
	for _, r := range d.records {
		keys += r.Keys
		typing += r.Typing
	}
	if typing <= 0 {
		return 0
	}
	return float64(keys) / typing.Minutes()
}

// MissRate は直近の窓のミス率を返す。
// 打ったかな (正解とミスの合計) のうちミスの割合。
func (d *Drill) MissRate() float64 {
	units, errors := 0, 0
	for _, r := range d.records {
		units += r.Units
		errors += r.Errors
	}
	if units+errors == 0 {
		return 0
	}
	return float64(errors) / float64(units+errors)
}

// WordResult は単語を打ち終えたときの判定。
type WordResult struct {
	Success  bool
	Duration time.Duration
	Errors   int
	Promoted bool
	// KanaAdded は昇格でかなが増えたことを示す。false の昇格は長さの解放。
	KanaAdded bool
	Demoted   bool
	// WeakUnit は降格の原因になったかな。
	WeakUnit string
}

// FinishWord は単語の結果を集計し、昇格・降格を反映する。
func (d *Drill) FinishWord(now time.Time) WordResult {
	out := WordResult{
		Duration: d.Elapsed(now),
		Errors:   d.wordErrors,
	}
	// 成功はミスの有無だけで決める。速さは窓の kpm で別に判定する
	out.Success = out.Errors == 0

	// 反応の猶予より速く打ち終えた語で速度が発散しないよう下限を置く
	typing := max(out.Duration-d.Cfg.ReactionBudget, 10*time.Millisecond)
	d.records = append(d.records, WordRecord{
		Success: out.Success,
		Units:   len(d.word),
		Keys:    d.wordKeys(),
		Errors:  out.Errors,
		Typing:  typing,
	})
	if len(d.records) > d.Cfg.WindowSize {
		d.records = d.records[len(d.records)-d.Cfg.WindowSize:]
	}
	if slices.Contains(d.word, d.Newest()) {
		d.newKanaWords++
	}

	// 正答率が下がったかなが出たら1つ降格する。
	// 覚えている最中のいちばん新しいかなは対象外。
	// 時間超過は打ち間違いではないので、ここには影響しない
	// （かなの統計は入力の正誤だけを記録している）。
	if d.Level > 1 {
		allowed := d.Allowed()
		for _, unit := range allowed[:len(allowed)-1] {
			s := d.Stats[unit]
			if s == nil || len(s.Recent) < d.Cfg.MinAttempts {
				continue
			}
			if s.RecentAccuracy() < d.Cfg.DemoteAccuracy {
				d.Level--
				out.Demoted = true
				out.WeakUnit = unit
				d.resetProgress()
				return out
			}
		}
	}

	// 窓が埋まっていて、打鍵速度とミス率の両方が基準を満たし、
	// 新出かなを含む語も十分に打っていたらレベルアップ
	if len(d.records) >= d.Cfg.WindowSize &&
		d.WindowKPM() >= d.Cfg.TargetKPM &&
		d.MissRate() <= d.Cfg.MaxMissRate &&
		d.newKanaWords >= d.GateTarget() &&
		d.Level < curriculum.MaxLevel() {
		before := len(d.Allowed())
		d.Level++
		out.Promoted = true
		out.KanaAdded = len(d.Allowed()) > before
		d.records = nil
		d.newKanaWords = 0
	}
	return out
}

// resetProgress は降格時に判定の記録を捨て、連鎖降格を防ぐ。
func (d *Drill) resetProgress() {
	d.records = nil
	d.newKanaWords = 0
	for _, st := range d.Stats {
		st.Recent = nil
	}
}

File: web/src/lib/drill.ts
// 練習セッションの進行と昇格・降格の判定。Go版 internal/drill の移植。
//
// 出題は1単語ずつ。単語が表示された瞬間から打ち終わるまでを計測する。
// 昇格は直近の窓の打鍵速度 (kpm) とミス率の両方が基準を満たしたとき。

import { count } from "./keys";
import { chordFor } from "./table";
import { maxLevel, stageFor, unitsFor, type Stage } from "./curriculum";

// 1つのかなの直近正答率を測る試行数。
const RECENT_WINDOW = 12;

export interface UnitStat {
  attempts: number;
  errors: number;
  recent: boolean[];
}

export function recentAccuracy(s: UnitStat): number {
  if (s.recent.length === 0) return 1;
  return s.recent.filter(Boolean).length / s.recent.length;
}

export interface DrillConfig {
  targetKPM: number; // 昇格に必要な打鍵速度 (keys per minute)
  maxMissRate: number; // 昇格できるミス率の上限
  windowSize: number; // 判定に使う直近の単語数
  minNewKanaWords: number; // 昇格までに打つ、新出かなを含む語の最低数
  requireBackspace: boolean; // ミス時にバックスペース修正が必要か
  demoteAccuracy: number; // これを下回ると降格するかなの直近正答率
  minAttempts: number; // 降格判定に必要な直近試行数
}

export const DEFAULT_DRILL_CONFIG: DrillConfig = {
  targetKPM: 120,
  maxMissRate: 0.05,
  windowSize: 100,
  minNewKanaWords: 50,
  requireBackspace: false,
  demoteAccuracy: 0.7,
  minAttempts: RECENT_WINDOW,
};

export interface WordRecord {
  success: boolean;
  units: number; // 正しく打ったかなの数
  keys: number; // 打鍵数 (同時押しは複数と数える)
  errors: number; // ミス入力の数
  typingMs: number; // 表示から打ち終わるまでの時間
}

export type InputResult =
  | "ignored"
  | "advance"
  | "error"
  | "wordDone"
  | "blocked" // バックスペース修正待ちの間のかな入力
  | "corrected"; // バックスペースで修正した

export interface WordResult {
  success: boolean;
  durationMs: number;
  errors: number;
  promoted: boolean;
  kanaAdded: boolean; // 昇格でかなが増えた。false の昇格は長さの解放
  demoted: boolean;
  weakUnit: string; // 降格の原因になったかな
}

export class Drill {
  level: number;
  stats: Record<string, UnitStat>;

  private word: string[] = [];
  private pos = 0;
  private wordErrors = 0;
  private pendingError = false; // バックスペース修正待ち
  private shownAt: number | null = null;
  private records: WordRecord[] = [];
  private newKanaWordCount = 0;

  constructor(
    public cfg: DrillConfig,
    level: number,
    stats: Record<string, UnitStat> = {},
  ) {
    this.level = Math.min(Math.max(level, 1), maxLevel());
    this.stats = stats;
    // 手で編集された進捗由来の null に耐える
    for (const k of Object.keys(this.stats)) {
      if (!this.stats[k]) delete this.stats[k];
    }
  }

  allowed(): string[] {
    return unitsFor(this.level);
  }

  stage(): Stage {
    return stageFor(this.level);
  }

  newest(): string {
    const a = this.allowed();
    return a[a.length - 1];
  }

  newKanaWords(): number {
    return this.newKanaWordCount;
  }

  // 最初の5文字の段階では全部が新出なので、ゲートはかけない。
  // それ以外は設定値そのまま (自動で緩めたりはしない)。
  gateTarget(): number {
    if (this.allowed().length === unitsFor(1).length) return 0;
    return this.cfg.minNewKanaWords;
  }

  progress(): { records: WordRecord[]; newKanaWords: number } {
    return { records: this.records, newKanaWords: this.newKanaWordCount };
  }

  setProgress(records: WordRecord[], newKanaWords: number): void {
    this.records = records.slice(-this.cfg.windowSize);
    this.newKanaWordCount = Math.max(newKanaWords, 0);
  }

  startWord(units: string[], nowMs: number): void {
    this.word = units;
    this.pos = 0;
    this.wordErrors = 0;
    this.pendingError = false;
    this.shownAt = nowMs;
  }

  currentWord(): string[] {
    return this.word;
  }

  currentPos(): number {
    return this.pos;
  }

  currentErrors(): number {
    return this.wordErrors;
  }

  expected(): string {
    return this.pos < this.word.length ? this.word[this.pos] : "";
  }

  // バックスペース修正待ちかどうか。
  needsBackspace(): boolean {
    return this.pendingError;
  }

  elapsedMs(nowMs: number): number {
    return this.shownAt === null ? 0 : nowMs - this.shownAt;
  }

  private wordKeys(): number {
    let keys = 0;
    for (const u of this.word) {
      const chord = chordFor(u);
      keys += chord === undefined ? 1 : count(chord);
    }
    return keys;
  }

  input(text: string): InputResult {
    if (this.pos >= this.word.length) return "ignored";

    // 修正必須モードでは、ミスの後にバックスペースを打つまで進めない
    if (this.pendingError) {
      if (text === "\b") {
        this.pendingError = false;
        return "corrected";
      }
      if (text === " " || text === "\n" || text === "") return "ignored";
      return "blocked";
    }

    if (text === " " || text === "\b" || text === "\n" || text === "") {
      return "ignored";
    }
    const expected = this.word[this.pos];
    if (text === expected) {
      this.record(expected, true);
      this.pos++;
      return this.pos >= this.word.length ? "wordDone" : "advance";
    }
    this.record(expected, false);
    this.wordErrors++;
    if (this.cfg.requireBackspace) {
      this.pendingError = true;
    }
    return "error";
  }

  private stat(unit: string): UnitStat {
    let s = this.stats[unit];
    if (!s) {
      s = { attempts: 0, errors: 0, recent: [] };
      this.stats[unit] = s;
    }
    return s;
  }

  private record(unit: string, ok: boolean): void {
    const s = this.stat(unit);
    s.attempts++;
    if (!ok) s.errors++;
    s.recent.push(ok);
    if (s.recent.length > RECENT_WINDOW) {
      s.recent = s.recent.slice(-RECENT_WINDOW);
    }
  }

  successCount(): { successes: number; total: number } {
    return {
      successes: this.records.filter((r) => r.success).length,
      total: this.records.length,
    };
  }

  // 直近の窓の打鍵速度 (keys per minute)。
  // 表示から打ち終わるまでの素の時間で計算する。
  windowKPM(): number {
    let keys = 0;
    let typingMs = 0;
    for (const r of this.records) {
      keys += r.keys;
      typingMs += r.typingMs;
    }
    if (typingMs <= 0) return 0;
    return keys / (typingMs / 60000);
  }

  // 直近の窓のミス率。打ったかな (正解とミスの合計) のうちミスの割合。
  missRate(): number {
    let units = 0;
    let errors = 0;
    for (const r of this.records) {
      units += r.units;
      errors += r.errors;
    }
    if (units + errors === 0) return 0;
    return errors / (units + errors);
  }

  finishWord(nowMs: number): WordResult {
    const out: WordResult = {
      success: this.wordErrors === 0,
      durationMs: this.elapsedMs(nowMs),
      errors: this.wordErrors,
      promoted: false,
      kanaAdded: false,
      demoted: false,
      weakUnit: "",
    };

    // 一瞬で打ち終えた語で速度が発散しないよう下限を置く。
    // ms は整数に丸める (保存と共有URLを短くするため)
    const typingMs = Math.max(Math.round(out.durationMs), 10);
    this.records.push({
      success: out.success,
      units: this.word.length,
      keys: this.wordKeys(),
      errors: out.errors,
      typingMs,
    });
    this.records = this.records.slice(-this.cfg.windowSize);
    if (this.word.includes(this.newest())) {
      this.newKanaWordCount++;
    }

    // 正答率が下がったかなが出たら1つ降格する。
    // 覚えている最中のいちばん新しいかなは対象外。
    if (this.level > 1) {
      const allowed = this.allowed();
      for (const unit of allowed.slice(0, -1)) {
        const s = this.stats[unit];
        if (!s || s.recent.length < this.cfg.minAttempts) continue;
        if (recentAccuracy(s) < this.cfg.demoteAccuracy) {
          this.level--;
          out.demoted = true;
          out.weakUnit = unit;
          this.resetProgress();
          return out;
        }
      }
    }

    // 窓が埋まっていて、打鍵速度とミス率の両方が基準を満たし、
    // 新出かなを含む語も十分に打っていたらレベルアップ
    if (
      this.records.length >= this.cfg.windowSize &&
      this.windowKPM() >= this.cfg.targetKPM &&
      this.missRate() <= this.cfg.maxMissRate &&
      this.newKanaWordCount >= this.gateTarget() &&
      this.level < maxLevel()
    ) {
      const before = this.allowed().length;
      this.level++;
      out.promoted = true;
      out.kanaAdded = this.allowed().length > before;
      this.records = [];
      this.newKanaWordCount = 0;
    }
    return out;
  }

  private resetProgress(): void {
    this.records = [];
    this.newKanaWordCount = 0;
    for (const s of Object.values(this.stats)) {
      s.recent = [];
    }
  }
}
